torch/lib/c10d/reducer.cpp

```diff
@@ -1286,6 +1305,7 @@ void Reducer::finalize_backward() {
   // See Note [Skip allreducing local_used_maps_dev]
   if (find_unused_parameters_) {
     // Reset unused parameter accounting.
+    // See Note [local_used_maps_ -> local_used_maps_dev copying]
     for (auto& local_used : local_used_maps_) {
       local_used.fill_(0);
     }
```

Comment:
we can move this reset to be after line 1321, to make sure local_used_maps_ is reset after cuda operations are done (possibly need to block copy from local_used_maps_ to local_used_maps_dev_ so that host and device are synced?).   Will this solve the issue as well?

Reply:
You mean, move it after `local_used_work_->wait()`?
interesting idea. If wait() blocks the host, you're right. But I think wait() only tells the current stream to wait on the nccl allreduce stream (in other words, I think it only affects the stream ordering of ops) and therefore won't save us from the host vs stream race condition.
let me double check what wait() does.

Reply:
Afaict, as long as the [NCCL_BLOCKING_WAIT](https://github.com/pytorch/pytorch/blob/8b5b7fa83d0a79860a8fe93206ea32121e63a81e/torch/lib/c10d/ProcessGroupNCCL.cpp#L411) environment variable is not set, [`wait()` calls `synchronizeInternal(timeout/*defaults to kNoTimeout*/)`](https://github.com/pytorch/pytorch/blob/8b5b7fa83d0a79860a8fe93206ea32121e63a81e/torch/lib/c10d/ProcessGroupNCCL.cpp#L381-L382), which tells the [current stream to sync on internal nccl stream(s)](https://github.com/pytorch/pytorch/blob/8b5b7fa83d0a79860a8fe93206ea32121e63a81e/torch/lib/c10d/ProcessGroupNCCL.cpp#L306-L317). In other words I don't think we can rely on wait() blocking the host.

Reply:
thanks! I'm wondering whether we've tried to add "local_used_maps_[i].copy_(local_used_maps_dev_[i]);" blocking copy after "local_used_work_->wait();" in line 1311? @mrshenli , did it have performance impact before? if we add this blocking copy there, the issue will be resolved. 
 
I guess the performance impact should be small here? As local_used_work_ has been kicked off for a while, and in any way all ranks needs to be synced before next iteration starts.

Reply:
I think my solution is a bit better, because in the case where all params are used, it never blocks.

Reply:
+1 to @mcarilli's comment. The current implementation will need `local_used_maps_` being update to detect `global_unused`, which comes before the `wait()` call. 

Curious, if the concern is `local_used.fill_(0)` occurs too early in `finalize_backward` before the CUDA async copy reads the value, will it work if we move `local_used.fill_(0)` to `prepare_for_backward` to zero out the tensor in the next forward pass?

Reply:
The main concern I had is the overhead of `pin_memory`.

Reply:
> will it work if we move local_used.fill_(0) to prepare_for_backward to zero out the tensor in the next forward pass?

That would make the race condition rarer, but still possible, if external user code doesn't sync between here and prepare_for_backward. The stream might have a large backlog of work, so host activity might run far ahead as long as no syncs occur (which as you both know is great for performance/utilization, we just have to avoid race conditions).

> The main concern I had is the overhead of pin_memory.
see https://github.com/pytorch/pytorch/pull/53160#discussion_r587917169.

Reply:
I'm proposing to another solution is to add "local_used_maps_[i].copy_(local_used_maps_dev_[i]);" blocking copy after "local_used_work_->wait();" in line 1311 and move local_used.fill_(0) after it.   This will sync host and device for the rank where all params are used.   But it may have perf impact as well. 

Let me benchmark this for Bert and Resnet as well. Will compare results with @mcarilli's solution here.

Reply:
That would work, but I think my approach will turn out alright perfwise. (Also, if you plan to use `local_used_maps_[i].copy_(local_used_maps_dev_[i]);` for its blocking side effect alone, you may as well just call cudaStreamSynchronize.)